### PR TITLE
 IO-spec update for link check logic

### PR DESCRIPTION
**Pull request by mlimbeck**
- Updated IO-spec for link check logic
- proof of link check logic in `process()`

TODO:
- [ ] prove assumption in `validateEgressID()`
- [x] prove assumption in `validateSrcDstIA()`

### CODE REVIEW

File: router/dataplane.go [hunk 1]
```diff
@@ -2198,10 +2202,13 @@ func (p *scionPacketProcessor) validateTransitUnderlaySrc( /*@ ghost ub []byte @
 // @ ensures   reserr != nil ==> reserr.ErrorMem()
 // contracts for IO-spec
 // @ requires  dp.Valid()
+// @ requires  p.d.DpAgreesWithSpec(dp)
 // @ requires  len(oldPkt.CurrSeg.Future) > 0
 // @ requires  p.EqAbsHopField(oldPkt)
 // @ requires  p.EqAbsInfoField(oldPkt)
 // @ requires  p.segmentChange ==> oldPkt.RightSeg != none[io.IO_seg2] && len(get(oldPkt.RightSeg).Past) > 0
+// @ requires  !p.segmentChange ==> AbsValidateIngressIDConstraint(oldPkt, path.ifsToIO_ifs(p.ingressID))
+// @ requires  p.segmentChange ==> AbsValidateIngressIDConstraintXover(oldPkt, path.ifsToIO_ifs(p.ingressID))
```
Comment (jcp19) [inline router/dataplane.go:2211]: nitpick: break these long lines (I will do this myself when I format the code in the basis PR)

File: router/io-spec-abstract-transitions.gobra [hunk 2]
```diff
@@ -50,6 +50,17 @@ pure func AbsValidateIngressIDConstraint(pkt io.IO_pkt2, ingressID option[io.IO_
 			ingressID == (currseg.ConsDir ? currseg.Future[0].InIF2 : currseg.Future[0].EgIF2)
 }
 
+ghost
+opaque
+requires pkt.RightSeg != none[io.IO_seg2]
```
Comment (jcp19) [inline router/io-spec-abstract-transitions.gobra:55]: if we require this to not be `none`, why don't we pass an `io.IO_ifs` instead of an `option`?

**Review by jcp19 — APPROVED** [review]

**Review by Dspil — APPROVED** [review]

File: pkg/slayers/path/scion/raw.go [hunk 3]
```diff
@@ -481,12 +481,13 @@ func (s *Raw) IsPenultimateHop( /*@ ghost ubuf []byte @*/ ) bool {
...
-// @ preserves acc(s.Mem(ubuf), R20)
+// @ preserves acc(s.Mem(ubuf), R40)
+// @ ensures res == s.IsLastHopSpec(ubuf)
 // @ decreases
-func (s *Raw) IsLastHop( /*@ ghost ubuf []byte @*/ ) bool {
-	//@ unfold acc(s.Mem(ubuf), R20)
-	//@ defer  fold acc(s.Mem(ubuf), R20)
-	//@ unfold acc(s.Base.Mem(), R20)
-	//@ defer  fold acc(s.Base.Mem(), R20)
+func (s *Raw) IsLastHop( /*@ ghost ubuf []byte @*/ ) /*@ (res @*/ bool /* @ ) @*/ {
```
Comment (jcp19) [inline pkg/slayers/path/scion/raw.go:487]:
```suggestion
func (s *Raw) IsLastHop( /*@ ghost ubuf []byte @*/ ) (res bool) {
```

File: router/io-spec-non-proven-lemmas.gobra [hunk 4]
```diff
@@ -74,39 +74,33 @@ func absPktFutureLemma(dp io.DataPlaneSpec, raw []byte) {
...
+requires  len(get(oldPkt.LeftSeg).Future) > 0
+requires  len(get(oldPkt.LeftSeg).History) == 0
+requires  AbsValidateIngressIDConstraint(oldPkt, ingressID)
+requires  newPkt == AbsDoXover(oldPkt)
+ensures   AbsValidateIngressIDConstraintXover(newPkt, ingressID)
+decreases
+func AbsValidateIngressIDXoverLemma(oldPkt io.IO_pkt2, newPkt io.IO_pkt2, ingressID option[io.IO_ifs]) {
+	reveal AbsValidateIngressIDConstraint(oldPkt, ingressID)
+	reveal AbsDoXover(oldPkt)
+	reveal AbsValidateIngressIDConstraintXover(newPkt, ingressID)
```
Comment (jcp19) [inline router/io-spec-non-proven-lemmas.gobra:101]: the lemmas that are currently proven should be moved to a different file (we can create one called `bytes-pkts-lemmas.gobra`)

**Review by jcp19 — APPROVED** [review]

**Pull request merged**